File: 516-longest-palindromic-subsequence/516-longest-palindromic-subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        int n = s.length();
        vector<vector<int>>dp(n+1,vector<int>(n+1));
        for(int i=1;i<=n;i++) dp[i][i] = 1;
        for(int l=2;l<=n;l++){
            for(int i=1;i<=n-l+1;i++){
                if(s[i-1]==s[i+l-2]){
                    dp[i][i+l-1] = dp[i+1][i+l-2]+2;
                }else{
                    dp[i][i+l-1] = max(dp[i+1][i+l-1],dp[i][i+l-2]);
                }
            }
        }
        return dp[1][n];
    }
};

File: 516-longest-palindromic-subsequence/516-longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        s2 = s[::-1]
        dp = []
        for i in range(0,n+1):
            temp = []
            for j in range(0,n+1):
                temp.append(0)
            dp.append(temp)
        for i in range(1,n+1):
            for j in range(1,n+1):
                if s[i-1]==s2[j-1]:
                    dp[i][j] = 1+dp[i-1][j-1]
                else:
                    dp[i][j] = max(dp[i-1][j],dp[i][j-1])
        return dp[n][n]
        